modules/etcd/etcd.go

```diff
@@ -0,0 +1,243 @@
+	// only the first node creates the cluster
+	if settings.currentNode == 0 {
+		for i := 1; i < len(settings.nodeNames); i++ {
+			// move to the next node
+			clusterNode, err := Run(ctx, req.Image, append(clusterOpts, withCurrentNode(i))...)
+			if err != nil {
+				// return the parent cluster node and the error, so the caller can clean up.
+				return c, fmt.Errorf("run cluster node: %w", err)
+			}
+
+			c.childNodes = append(c.childNodes, clusterNode)
+		}
+	}
+
+	c.ClusterToken = settings.clusterToken
+
+	return c, nil
+}
+
+// configureCluster configures the cluster settings, ensuring that the cluster is properly configured with the necessary network and options,
+// avoiding duplicate application of options to be passed to the successive nodes.
+func configureCluster(ctx context.Context, settings *options, opts []testcontainers.ContainerCustomizer) ([]testcontainers.ContainerCustomizer, error) {
+	var clusterOpts []testcontainers.ContainerCustomizer
+	if len(settings.nodeNames) <= 1 {
+		return clusterOpts, nil
+	}
+
+	// pass cluster options to each node
+	etcdOpts := []Option{}
+	for _, opt := range opts {
+		// if the option is of type Option, it won't be applied to the settings
+		// this prevents the same option from being applied multiple times (e.g. updating the current node)
+		if apply, ok := opt.(Option); ok {
+			etcdOpts = append(etcdOpts, apply)
+		} else {
+			clusterOpts = append(clusterOpts, opt)
+		}
+	}
+
+	if settings.clusterNetwork == nil { // the first time the network is created
+		newNetwork, err := tcnetwork.New(ctx)
+		if err != nil {
+			return clusterOpts, fmt.Errorf("new network: %w", err)
+		}
+
+		// set the network for the first node
+		settings.clusterNetwork = newNetwork
+
+		clusterOpts = append(clusterOpts, withClusterNetwork(newNetwork)) // save the network for the next nodes
+	}
+
+	// we finally need to re-apply all the etcd-specific options
+	clusterOpts = append(clusterOpts, withClusterOptions(etcdOpts))
+
+	return clusterOpts, nil
+}
+
+// configureCMD configures the etcd command line arguments, based on the settings provided,
+// in order to create a cluster or a single-node instance.
+func configureCMD(settings options) []string {
+	cmds := []string{"etcd"}
+
+	if len(settings.nodeNames) < 1 {
+		cmds = append(cmds, "--name=default")
+	} else {
+		clusterCmds := []string{
+			"--name=" + settings.nodeNames[settings.currentNode],
+			"--initial-advertise-peer-urls=" + scheme + "://" + settings.nodeNames[settings.currentNode] + ":" + peerPort,
+			"--advertise-client-urls=" + scheme + "://" + settings.nodeNames[settings.currentNode] + ":" + clientPort,
+			"--listen-peer-urls=" + scheme + "://0.0.0.0:" + peerPort,
+			"--listen-client-urls=" + scheme + "://0.0.0.0:" + clientPort,
+			"--initial-cluster-state=new",
+		}
+
+		clusterStateValues := []string{}
+		for _, node := range settings.nodeNames {
+			clusterStateValues = append(clusterStateValues, node+"="+scheme+"://"+node+":"+peerPort)
+		}
```

Comment:
suggestion: this could be preallocated
```suggestion
		clusterStateValues := make([]string, len(settings.nodeNames))
		for i, node := range settings.nodeNames {
			clusterStateValues[i] = node+"="+scheme+"://"+node+":"+peerPort)
		}
```